### PR TITLE
Refactor the DataFrame APIs to use the latest versions where possible

### PR DESCRIPTION
**Pull request by pgovind**
This can be reviewed now.

Old:
This is WIP. I just have it up here for initial review. 4 things need to happen before this can come out of WIP:
1) https://github.com/dotnet/corefxlab/pull/2885 needs to go in - Done
2) We need to make a 0.3.0 for Microsoft.Data.Analysis - Done
3) 0.3.0 has to go on to Nuget officially - Done
3) The first commit in this PR has to be reverted and the MDA dependency in Spark.csproj needs to be updated to 0.3.0 - Done

 @eerhardt

### CODE REVIEW

File: src/csharp/Microsoft.Spark.sln [hunk 1]
```diff
@@ -33,6 +33,8 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Spark.Extensions.
 EndProject
 Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Spark.Extensions.Delta.E2ETest", "Extensions\Microsoft.Spark.Extensions.Delta.E2ETest\Microsoft.Spark.Extensions.Delta.E2ETest.csproj", "{206E16CA-ED59-4F5E-8EA1-9BB7BEEACB63}"
 EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "Microsoft.Data.Analysis", "..\..\..\corefxlab\src\Microsoft.Data.Analysis\Microsoft.Data.Analysis.csproj", "{972FDC86-4029-4342-829E-51C2E0B170F6}"
```
Comment (pgovind) [inline src/csharp/Microsoft.Spark.sln]: Needs to be reverse once steps 1 to 3 are complete

File: src/csharp/Microsoft.Spark/Microsoft.Spark.csproj [hunk 2]
```diff
@@ -26,18 +26,19 @@
   <ItemGroup>
```
Comment (pgovind) [inline src/csharp/Microsoft.Spark/Microsoft.Spark.csproj]: Needs to be reverted once steps 1 to 3 are done

File: src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs [hunk 3]
```diff
@@ -226,7 +226,7 @@ public void TestVectorUdf()
         [Fact]
         public void TestDataFrameVectorUdf()
         {
-            Func<PrimitiveDataFrameColumn<int>, ArrowStringDataFrameColumn, ArrowStringDataFrameColumn> udf1Func =
+            Func<Int32DataFrameColumn, ArrowStringDataFrameColumn, ArrowStringDataFrameColumn> udf1Func =
                 (ages, names) =>
                 {
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs]: Will https://github.com/dotnet/spark/pull/277#discussion_r387794908 be addressed as well?

Comment (pgovind) [reply src/csharp/Microsoft.Spark.E2ETest/IpcTests/Sql/DataFrameTests.cs]: Yup. https://github.com/dotnet/corefxlab/pull/2889 is up to address this.

File: src/csharp/Microsoft.Spark.UnitTest/WorkerFunctionTests.cs [hunk 4]
```diff
@@ -208,44 +208,42 @@ public void TestChainingArrowWorkerFunction()
...
+                new DataFrameUdfWrapper<Int32DataFrameColumn, ArrowStringDataFrameColumn, ArrowStringDataFrameColumn>(
                     (numbers, strings) =>
                     {
-                        var stringColumn = (StringArray)ToArrowArray(
-                             Enumerable.Range(0, (int)strings.Length)
-                                .Select(i => $"{strings[i]}:{numbers[i]}")
-                                .ToArray());
-                        return ToArrowStringDataFrameColumn(stringColumn);
+                        long i = 0;
+                        return strings.Apply((string cur) =>
```
Comment (eerhardt) [inline src/csharp/Microsoft.Spark.UnitTest/WorkerFunctionTests.cs]: (nit) this could get even simpler:

```
long i = 0;
return strings.Apply(cur => $"{cur}:{numbers[i++]}");
```

**Comment by pgovind** [comment]
Submitted https://github.com/dotnet/corefxlab/pull/2890 to bump the version. All the tests pass in CI.

File: NuGet.config [hunk 5]
```diff
@@ -1,4 +1,4 @@
-<?xml version="1.0" encoding="utf-8"?>
+﻿<?xml version="1.0" encoding="utf-8"?>
```
Comment (pgovind) [inline NuGet.config]: Revert this change

**Comment by pgovind** [comment]
This should be ready to go in, pending approval of course.

File: eng/Versions.props [hunk 6]
```diff
@@ -6,6 +6,7 @@
     <RestoreSources>
       $(RestoreSources);
       https://api.nuget.org/v3/index.json;
+      https://pkgs.dev.azure.com/dnceng/public/_packaging/dotnet-experimental/nuget/v3/index.json;
```
Comment (pgovind) [inline eng/Versions.props]: I'm assuming its's ok to leave this in? That way the next time we need to update MDA, we don't have to search for the feed?

Comment (imback82) [reply eng/Versions.props]: I would remove it unless absolutely necessary. :)

File: examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs [hunk 7]
```diff
@@ -62,9 +62,9 @@ private static FxDataFrame CountCharacters(FxDataFrame dataFrame)
         {
             int characterCount = 0;
 
-            var characterCountColumn = new PrimitiveDataFrameColumn<int>("nameCharCount");
-            var ageColumn = new PrimitiveDataFrameColumn<int>("age");
-            ArrowStringDataFrameColumn nameColumn = dataFrame["name"] as ArrowStringDataFrameColumn;
+            var characterCountColumn = new Int32DataFrameColumn("nameCharCount");
+            var ageColumn = new Int32DataFrameColumn("age");
+            ArrowStringDataFrameColumn nameColumn = dataFrame.Columns.GetArrowStringColumn("name");
```
Comment (imback82) [inline examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs]: Question on the API. Does the user need to know about `Arrow` here? Can I just get a `String` column?

File: src/csharp/Microsoft.Spark.UnitTest/WorkerFunctionTests.cs [hunk 8]
```diff
@@ -208,44 +208,33 @@ public void TestChainingArrowWorkerFunction()
...
+                new DataFrameUdfWrapper<Int32DataFrameColumn, ArrowStringDataFrameColumn, ArrowStringDataFrameColumn>(
                     (numbers, strings) =>
                     {
-                        var stringColumn = (StringArray)ToArrowArray(
-                             Enumerable.Range(0, (int)strings.Length)
-                                .Select(i => $"{strings[i]}:{numbers[i]}")
-                                .ToArray());
-                        return ToArrowStringDataFrameColumn(stringColumn);
+                        long i = 0;
+                        return strings.Apply(cur => $"{cur}:{numbers[i++]}");
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.UnitTest/WorkerFunctionTests.cs]: Nice!

File: src/csharp/Microsoft.Spark.Worker.UnitTest/CommandExecutorTests.cs [hunk 9]
```diff
@@ -309,11 +309,7 @@ public async Task TestDataFrameSqlCommandExecutorWithSingleCommand()
             var udfWrapper = new Sql.DataFrameUdfWrapper<ArrowStringDataFrameColumn, ArrowStringDataFrameColumn>(
                 (strings) =>
                 {
-                    var stringArray = (StringArray)ToArrowArray(
-                        Enumerable.Range(0, (int)strings.Length)
-                            .Select(i => $"udf: {strings[i]}")
-                            .ToArray());
-                    return ToArrowStringDataFrameColumn(stringArray);
+                    return strings.Apply(cur => $"udf: {cur}");
```
Comment (imback82) [inline src/csharp/Microsoft.Spark.Worker.UnitTest/CommandExecutorTests.cs]: just `(strings) => strings.Apply(cur => $"udf: {cur}")`?

File: examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs (hunk 7 above)
Comment (pgovind) [reply examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs]: It's because we have 2 string columns: a `StringDataFrameColumn` and an `ArrowStringDataFrameColumn`. Maybe we should've named them `UTF16DataFrameColumn` and `UTF8DataFrameColumn`! `GetArrowStringColumn` returns the `ArrowStringDataFrameColumn` while `GetStringColumn` would return `StringDataFrameColumn`.

File: src/csharp/Microsoft.Spark.Worker.UnitTest/CommandExecutorTests.cs (hunk 9 above)
Comment (pgovind) [reply src/csharp/Microsoft.Spark.Worker.UnitTest/CommandExecutorTests.cs]: Yup. Fixed in other places too.

File: examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs (hunk 7 above)
Comment (imback82) [reply examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs]: Got it, thanks. +1 for `UTF16DataFrameColumn` and `UTF8DataFrameColumn`

Comment (pgovind) [reply examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs]: cc @eerhardt on the column names. Maybe something to consider for the future?

Comment (eerhardt) [reply examples/Microsoft.Spark.CSharp.Examples/Sql/Batch/VectorDataFrameUdfs.cs]: I think just `StringDataFrameColumn` should stay `StringDataFrameColumn`, since `string` is the common .NET type.

But I could imagine renaming `ArrowStringDataFrameColumn` to `Utf8StringDataFrameColumn`.
